enumerables.rb

```diff
@@ -0,0 +1,293 @@
+
+    # Not there a block or an argument?
+    elsif arg.nil?
+      my_each { |elmt| counter += 1 if elmt != false && !elmt.nil? }
+
+    # My_all? will be used to find out if the collection is made of a specific object
+    else
+      my_each { |elmt| counter += 1 if elmt == arg }
+    end
+    counter == to_a.length
+  end
+end
+
+module Enumerable
+  # 5.- ---- my_any? Method ----
+
+  def my_any?(arg = nil)
+    counter = 0
+
+    #  Was a block given to the method?
+    if block_given?
+      my_each { |elmt| counter += 1 if yield(elmt) == true }
+
+    #  If there's no block, is the given argument the name of a data type like Integer, String, Array etc.?
+    elsif arg.is_a? Class
+      my_each { |elmt| counter += 1 if elmt.class == arg or elmt.is_a? arg }
+
+    # Does any of the elements match the regular expression?
+    elsif arg.class == Regexp
+      arr_str = map { |i| i.to_s }
+      arr_str.my_each { |elmt| counter += 1 if elmt.match?(arg) }
+
+    # Not there a block or an argument?
+    elsif arg.nil?
+      my_each { |elmt| counter += 1 if elmt != false && !elmt.nil? }
+
+    # My_any? will be used to find out if the collection is made of a specific object
+    else
+      my_each { |elmt| counter += 1 if elmt == arg }
+    end
+    counter.positive?
+  end
+end
+
+module Enumerable
+  # 4.- ---- my_none? Method ----
+
+  def my_none?(arg = nil)
+    counter = 0
+
+    #  Was a block given to the method?
+    if block_given?
+      my_each { |elmt| counter += 1 if yield(elmt) == true }
+
+    #  If there's no block, is the given argument the name of a data type like Integer, String, Array etc.?
+    elsif arg.is_a? Class
+      my_each { |elmt| counter += 1 if elmt.class == arg or elmt.is_a? arg }
+
+    # Do none of the elements match the regular expression?
+    elsif arg.class == Regexp
+      arr_str = map { |i| i.to_s }
+      arr_str.my_each { |elmt| counter += 1 if elmt.match?(arg) }
+
+    # Not there a block or an argument?
+    elsif arg.nil?
+      my_each { |elmt| counter += 1 if elmt != false && !elmt.nil? }
+
+    # My_none? will be used to find out if the collection is not made of a specific object
+    else
+      my_each { |elmt| counter += 1 if elmt == arg }
+    end
+    counter.zero?
+  end
+end
+
+module Enumerable
+  # 7.- ---- my_count Method ----
+
+  def my_count(*arg)
+    count = 0
+
+    if block_given?
+      my_each do |elmt|
+        count += 1 if yield(elmt)
+      end
+    elsif arg.empty?
+      count = size
+    else
+      my_each do |elmt|
+        count += 1 if elmt == arg[0]
+      end
+    end
+    count
+  end
+
+  # 8.- ---- my_map Method ----
+
+  def my_map
+    return to_enum(:my_map) unless block_given?
+
+    new_arr = []
+
+    if self.class == Array or self.class == Range
+      my_each { |i| new_arr.push(yield(i)) }
+    elsif self.class == Hash
+      my_each { |key, value| new_arr.push(yield(key, value)) }
+    end
+    new_arr
+  end
+end
```

Comment:
* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 

- [ ] returns an Enumerator if no block is given
```ruby
[2,4,7,11].my_map should be equal to [2,4,7,11].map #<Enumerator: [2, 4, 7, 11]:map
```
<img width="682" alt="3" src="https://user-images.githubusercontent.com/58504655/98772955-b54d9c00-2409-11eb-8191-2d527d34f9d5.PNG">


- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

Reply:
Arslan Bisharat, hope you are great.
Is this a valid output when calling my_map method without any proc or block?

This is what I get when comparing #my_map with #map in Repl.it . I must admit still need to know more about when methods return Enumerators, it took me some time trying to figure out:

![image](https://user-images.githubusercontent.com/63531796/98873531-31fa6d80-243e-11eb-9301-6224a807039b.png)

(IDE output: ) 
![image](https://user-images.githubusercontent.com/63531796/98873590-51919600-243e-11eb-87e8-fb2867a06224.png)